lib/matplotlib/units.py

```diff
@@ -43,12 +43,80 @@ def default_units(x, axis):
 """
 
 from numbers import Number
+import inspect
 
 import numpy as np
 
 from matplotlib.cbook import iterable, safe_first_element
 
 
+def _accepts_units(convert_x, convert_y):
+    """
+    A decorator for functions and methods that accept units. The parameters
+    indicated in *convert_x* and *convert_y* are used to update the axis
+    unit information, are converted, and then handed on to the decorated
+    function.
+
+    Parameters
+    ----------
+    convert_x, convert_y : list
+        A list of integers or strings, indicating the arguments to be converted
+    """
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            axes = args[0]
+            # Bind the incoming arguments to the function signature
+            bound_args = inspect.signature(func).bind(*args, **kwargs)
+            # Get the original arguments - these will be modified later
+            arguments = bound_args.arguments
+            # Check for data kwarg
+            has_data = (('data' in arguments) and
+                        (arguments['data'] is not None))
+            if has_data:
+                data = arguments['data']
```

Comment:
Perhaps move the decorator under _preprocess_data, so that you don't have to reproduce its logic here?

Reply:
I tried that a while ago, and from what I remember it became quite complicated because `_preprocess_data` alters functions signatures in a funny way.

Reply:
I'm not quite sure, that it's a good idea for `_preprocess_data` to alter the function signature anyway. I would be happy explicitly writing a `data=None` in the signatures and only let `_preprocess_data` only work on that. Feels less magical and simplifies the `_preprocess_data` code.

Reply:
I'm not quite sure, that it's a good idea for `_preprocess_data` to alter the function signature anyway. I would be happy explicitly writing a `data=None` in the signatures and only let `_preprocess_data` only work on that. Feels less magical and simplifies the `_preprocess_data` code.

Reply:
I don't think that would work, since the idea is that if I do `ax.plot(dataframe['Distance'])` it will automatically get the index of the dataframe as the x-values. Anyway, if anyone wants to alter  the `_preprocess_data` decorator feel free to in another PR, but this one is for units 😉

Reply:
Separate topic: If I'm correct, the four previous lines can be simplified to
`data = arguments.get('data')`

and use `data is not None` instead of has_data in the following.